botocore/utils.py

```diff
@@ -260,70 +265,108 @@ def __init__(self, timeout=DEFAULT_METADATA_SERVICE_TIMEOUT,
             proxies=get_environ_proxies(self._url),
         )
 
-    def _get_request(self, url, timeout, num_attempts=1):
+    def retrieve_iam_role_credentials(self):
+        try:
+            role_name = self._get_iam_role()
+            credentials = self._get_credentials(role_name)
+            if self._contains_all_credential_fields(credentials):
+                return {
+                    'role_name': role_name,
+                    'access_key': credentials['AccessKeyId'],
+                    'secret_key': credentials['SecretAccessKey'],
+                    'token': credentials['Token'],
+                    'expiry_time': credentials['Expiration'],
+                }
+            else:
+                # IMDS can return a 200 response that has a JSON formatted
+                # error message (i.e. if ec2 is not trusted entity for the
+                # attached role). We do not necessarily want to retry for
+                # these and we also do not necessarily want to raise a key
+                # error. So at least log the problematic response and return
+                # an empty dictionary to signal that it was not able to
+                # retrieve credentials.
+                logger.debug('Response received %s is missing '
+                             'credentials.', credentials)
```

Comment:
I don't think we should be logging the body here.  We can't guarantee what's in the body.  It seems safer to call out specific cases we know about, e.g check for the specific schema for when ec2 is not a trusted entity for the attached role and print the response in that scenario.

Otherwise maybe print the key names but not the values or something similar.

Reply:
>> Check for the specific schema for when ec2 is not a trusted entity for the attached role and print the response in that scenario. Otherwise maybe print the key names but not the values or something similar.

I will probably do both: one log message for JSON errors and another that just prints out the missing keys.